File: nexus_core/embeddings.py
"""
NEXUS — Embeddings Client
===========================
Gemini text-embedding-004 via google-generativeai.
Simulation mode returns deterministic mock vectors.

Faithful port of the TypeScript embeddings from src/lib/embeddings.ts.
"""

from __future__ import annotations

import math
from typing import Optional

# ─── Module-level state ───────────────────────────────────────

_gemini_configured = False
_api_key: Optional[str] = None

_EMBEDDING_DIM = 768


# ─── Configuration ────────────────────────────────────────────

def configure_gemini(api_key: str) -> None:
    """Configure the Google Generative AI SDK with the given API key."""
    global _gemini_configured, _api_key  # noqa: PLW0603
    _api_key = api_key
    try:
        import google.generativeai as genai  # type: ignore
        genai.configure(api_key=api_key)
        _gemini_configured = True
    except ImportError:
        _gemini_configured = False


# ─── Real Embeddings ──────────────────────────────────────────

def get_embeddings(
    texts: list[str],
    model: str = "text-embedding-004",
) -> list[list[float]]:
    """
    Get real embeddings via the Gemini ``batchEmbedContents`` API.

    Falls back to mock embeddings on any failure.

    Args:
        texts: List of text strings to embed.
        model: Gemini embedding model name.

    Returns:
        List of embedding vectors (one per input text).
    """
    if not _gemini_configured or not _api_key:
        return get_mock_embeddings(texts)

    try:
        import google.generativeai as genai  # type: ignore

        results: list[list[float]] = []
        batch_size = 20

        for i in range(0, len(texts), batch_size):
            batch = texts[i : i + batch_size]
            try:
                response = genai.embed_content(
                    model=f"models/{model}",
                    content=batch,
                    task_type="retrieval_document",
                )
                results.extend(response["embedding"])
            except Exception:
                # Fallback to mock for this batch
                results.extend(get_mock_embeddings(batch))

        return results

    except ImportError:
        return get_mock_embeddings(texts)


def get_single_embedding(
    text: str,
    model: str = "text-embedding-004",
) -> list[float]:
    """
    Get a single embedding vector for a query string.

    Args:
        text:  The text to embed.
        model: Gemini embedding model name.

    Returns:
        A single embedding vector.
    """
    if not _gemini_configured or not _api_key:
        return get_mock_embeddings([text])[0]

    try:
        import google.generativeai as genai  # type: ignore

        response = genai.embed_content(
            model=f"models/{model}",
            content=text,
            task_type="retrieval_query",
        )
        return response["embedding"]

    except (ImportError, Exception):
        return get_mock_embeddings([text])[0]


# ─── Mock Embeddings (Simulation) ─────────────────────────────

def get_mock_embeddings(texts: list[str], dim: int = _EMBEDDING_DIM) -> list[list[float]]:
    """
    Generate deterministic pseudo-embeddings for simulation.

    Uses a simple hash of the text to produce a consistent unit
    vector.  The vectors are NOT semantically meaningful, but they
    allow the pipeline to run end-to-end without API calls.

    Args:
        texts: List of text strings.
        dim:   Embedding dimensionality (default 768).

    Returns:
        List of unit-normalised pseudo-embedding vectors.
    """
    results: list[list[float]] = []

    for text in texts:
        vector = [0.0] * dim

        # Compute a 32-bit hash from the text (same algorithm as TS)
        seed = 0
        for ch in text:
            seed = ((seed << 5) - seed + ord(ch)) & 0xFFFFFFFF

        # Simple LCG PRNG seeded by text hash
        state = abs(seed) or 42
        for i in range(dim):
            state = (state * 1103515245 + 12345) & 0x7FFFFFFF
            vector[i] = (state / 0x7FFFFFFF) * 2 - 1  # range [-1, 1]

        # Normalise to unit vector
        norm = math.sqrt(sum(v * v for v in vector))
        if norm > 0:
            vector = [v / norm for v in vector]

        results.append(vector)

    return results

File: src/lib/embeddings.ts
/* ═══════════════════════════════════════════════════════════
   NEXUS — Embeddings Client
   Gemini text-embedding-004 via /api/gemini route
   Simulation mode returns deterministic mock vectors
   ═══════════════════════════════════════════════════════════ */

const EMBEDDING_DIM = 768;

/**
 * Get real embeddings via Gemini API (server-side proxy).
 * Calls /api/gemini with mode="embed".
 */
export async function getGeminiEmbeddings(
  texts: string[],
  apiKey: string
): Promise<number[][]> {
  const res = await fetch('/api/gemini', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ mode: 'embed', apiKey, texts }),
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error((err as Record<string, string>).error || `Embedding API failed (${res.status})`);
  }

  const data = await res.json();
  return data.embeddings as number[][];
}

/**
 * Get a single embedding for a query.
 */
export async function getQueryEmbedding(
  query: string,
  apiKey: string
): Promise<number[]> {
  const embeddings = await getGeminiEmbeddings([query], apiKey);
  return embeddings[0];
}

/**
 * Simulation mode: generate deterministic pseudo-embeddings.
 * Uses a simple hash of the text to produce a consistent vector.
 * The vectors are NOT semantically meaningful, but they allow
 * the UI to demonstrate the full pipeline without API calls.
 */
export function getMockEmbeddings(texts: string[]): number[][] {
  return texts.map(text => {
    const vector = new Array(EMBEDDING_DIM);
    let seed = 0;
    for (let i = 0; i < text.length; i++) {
      seed = ((seed << 5) - seed + text.charCodeAt(i)) | 0;
    }
    // Simple PRNG seeded by text hash
    let state = Math.abs(seed) || 42;
    for (let i = 0; i < EMBEDDING_DIM; i++) {
      state = (state * 1103515245 + 12345) & 0x7fffffff;
      vector[i] = (state / 0x7fffffff) * 2 - 1; // range [-1, 1]
    }
    // Normalize to unit vector
    const norm = Math.sqrt(vector.reduce((s, v) => s + v * v, 0));
    return norm === 0 ? vector : vector.map(v => v / norm);
  });
}

/**
 * Batch embed chunks, respecting API rate limits.
 * Batches of 20 texts per API call, with 100ms delay between batches.
 */
export async function embedChunks(
  chunks: { id: string; content: string }[],
  apiKey: string,
  simulationMode: boolean = false,
  onProgress?: (done: number, total: number) => void
): Promise<Map<string, number[]>> {
  const result = new Map<string, number[]>();

  if (simulationMode) {
    // Simulation: generate all at once
    const texts = chunks.map(c => c.content);
    const embeddings = getMockEmbeddings(texts);
    chunks.forEach((chunk, i) => result.set(chunk.id, embeddings[i]));
    onProgress?.(chunks.length, chunks.length);
    return result;
  }

  // Real embeddings: batch of 20 per request
  const BATCH_SIZE = 20;
  const DELAY_MS = 100;

  for (let i = 0; i < chunks.length; i += BATCH_SIZE) {
    const batch = chunks.slice(i, i + BATCH_SIZE);
    const texts = batch.map(c => c.content);

    try {
      const embeddings = await getGeminiEmbeddings(texts, apiKey);
      batch.forEach((chunk, j) => result.set(chunk.id, embeddings[j]));
    } catch (e) {
      console.warn(`Embedding batch ${i / BATCH_SIZE + 1} failed:`, e);
      // Fall back to mock embeddings for this batch
      const mockEmbeddings = getMockEmbeddings(texts);
      batch.forEach((chunk, j) => result.set(chunk.id, mockEmbeddings[j]));
    }

    onProgress?.(Math.min(i + BATCH_SIZE, chunks.length), chunks.length);

    if (i + BATCH_SIZE < chunks.length) {
      await new Promise(resolve => setTimeout(resolve, DELAY_MS));
    }
  }

  return result;
}
